src/consensus.rs

```diff
@@ -292,19 +296,53 @@ impl<T: Proposition> Consensus<T> {
         Ok(None)
     }
 
+    pub fn detect_byzantine_voters(
+        &self,
+        signed_vote: &SignedVote<T>,
+    ) -> std::result::Result<(), BTreeMap<NodeId, Fault<T>>> {
+        let mut faults = BTreeMap::new();
+        if let Some(existing_vote) = self.votes.get(&signed_vote.voter) {
+            let fault = Fault::ChangedVote {
+                a: existing_vote.clone(),
+                b: signed_vote.clone(),
+            };
+
+            if let Ok(()) = fault.validate(&self.elders) {
+                faults.insert(signed_vote.voter, fault);
+            }
+        }
+
+        {
+            let fault = Fault::InvalidFault {
+                signed_vote: signed_vote.clone(),
+            };
+            if let Ok(()) = fault.validate(&self.elders) {
+                faults.insert(signed_vote.voter, fault);
+            }
+        }
+
+        // TODO: recursively check for faulty votes
+
+        if faults.is_empty() {
+            Ok(())
+        } else {
+            Err(faults)
+        }
+    }
+
     /// Validates a vote recursively all the way down to the proposition (T)
     /// Assumes those propositions are correct, they MUST be checked beforehand by the caller
     pub fn validate_signed_vote(&self, signed_vote: &SignedVote<T>) -> Result<()> {
-        self.verify_sig_share(&signed_vote.vote, signed_vote.voter, &signed_vote.sig)?;
+        signed_vote.validate_signature(&self.elders)?;
         self.validate_vote(&signed_vote.vote)?;
-        self.validate_vote_supersedes_existing_vote(signed_vote)?;
```

Comment:
Is this check done in fault detection now?

Reply:
It is, yeah, the `ChangedVote` fault checks that the vote supersedes the existing vote